src/overlay/TCPPeer.cpp

```diff
@@ -487,6 +500,14 @@ TCPPeer::startRead()
                 }
                 noteFullyReadBody(length);
                 recvMessage();
+                if (!hasReadingCapacity())
+                {
+                    // Break and wait until more capacity frees up
+                    CLOG_TRACE(Overlay, "Throttle reading for peer {}!",
+                               mApp.getConfig().toShortString(getPeerID()));
+                    mShouldScheduleRead = false;
```

Comment:
This part seems to be the key, and I am happy to see it! I am sorry the rest of the review seems so critical -- I am I guess somewhat curious to see if you can get away with something smaller that just uses this, the message trackers, and some adjustments / improvements to the write-queue load shedding on the sending side.